src/Components/Web/test/Forms/ValidationSummaryTest.cs

```diff
@@ -0,0 +1,115 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+#nullable enable
+
+using Microsoft.AspNetCore.Components.Rendering;
+using Microsoft.AspNetCore.Components.RenderTree;
+using Microsoft.AspNetCore.Components.Test.Helpers;
+
+namespace Microsoft.AspNetCore.Components.Forms;
+
+public class ValidationSummaryTest
+{
+    [Theory]
+    [InlineData(null)]
+    [InlineData("")]
+    [InlineData(" ")]
+    public async Task ValidationSummary_HasDefaultCssClass_WhenNoCssClassIsProvided(string? cssClass)
+    {
+        // Arrange
+        var model = new TestModel();
+        var editContext = new EditContext(model);
+        var rootComponent = new TestValidationSummaryHostComponent
+        {
+            EditContext = editContext,
+            CssClass = cssClass,
+        };
+
+        var fieldIdentifier = FieldIdentifier.Create(() => model.StringProperty);
+        var messageStore = new ValidationMessageStore(editContext);
+        messageStore.Add(fieldIdentifier, "Test error message");
+
+        // Act
+        var renderer = new TestRenderer();
+        var rootComponentId = renderer.AssignRootComponentId(rootComponent);
+        await renderer.RenderRootComponentAsync(rootComponentId);
+        var batch = renderer.Batches.Single();
+        var ulFrame = batch.ReferenceFrames.Single(x => x.ElementName == "ul");
+        var ulFrameIndex = Array.IndexOf(batch.ReferenceFrames, ulFrame);
+        var ulFrameSubtree = batch.ReferenceFrames.Skip(ulFrameIndex).Take(ulFrame.ElementSubtreeLength);
+
+        var ulClassFrame = ulFrameSubtree.FirstOrDefault(x =>
```

Comment:
I'm not used to writing tests for components without bUnit. Is this the easiest way to find the resulting `class` attribute for the generated `ul` element?

Reply:
Typically we don't write unit tests for this kind of component rendering. Instead we write end-to-end (E2E) tests that rendering into a real browser and are automated with Selenium.

If you have a look at `src\Components\test\E2ETest\Tests\FormsTest.cs` and `TypicalValidationComponent.razor` you'll see some examples. It should be reasonably straightforward to add a further case following this pattern to show that `ValidationSummary` renders a supplied class name. These E2E tests can be executed as normal tests via `dotnet test` or the VS test window, but if you run the whole suite of them it can take a while, like 15 minutes or more.

Reply:
Sounds good. I switched this to an E2E test. Had a hell of a time getting selenium to run because the web driver requires v103 of chrome but the latest version of chrome is v105 and there is no way to download an old version of chrome. Ended up having to find a build of chromium for v103. You may want to add this info to the build docs. I remember now why I switched away from selenium and started using playwright instead :)